src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 	uint32 rangeTableId = 1;
 	Oid insertRelationId = insertRte->relid;
 	Var *insertPartitionColumn = PartitionColumn(insertRelationId, rangeTableId);
-	bool partitionColumnsMatch = false;
 	Query *subquery = subqueryRte->subquery;
+	bool targetTableHasPartitionColumn = false;
 
 	foreach(targetEntryCell, query->targetList)
 	{
 		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		List *insertTargetEntryColumnList = pull_var_clause_default((Node *) targetEntry);
+		int insertTargetEntryColumnCount = list_length(insertTargetEntryColumnList);
+		Var *insertVar = NULL;
+		AttrNumber originalAttrNo = InvalidAttrNumber;
+		TargetEntry *subqeryTargetEntry = NULL;
 
-		if (IsA(targetEntry->expr, Var))
+		/*
+		 * We only consider target entries that include a single column. Note that this
+		 * is slightly different than directly checking the whether the targetEntry->expr
+		 * is a var since the var could be wrapped into an implicit/explicit casting.
+		 */
+		if (insertTargetEntryColumnCount != 1)
 		{
-			Var *insertVar = (Var *) targetEntry->expr;
-			AttrNumber originalAttrNo = get_attnum(insertRelationId,
-												   targetEntry->resname);
-			TargetEntry *subqeryTargetEntry = NULL;
+			continue;
+		}
 
-			if (originalAttrNo != insertPartitionColumn->varattno)
-			{
-				continue;
-			}
+		insertVar = (Var *) linitial(insertTargetEntryColumnList);
+		originalAttrNo = get_attnum(insertRelationId,
```

Comment:
Is this call necessary? How does this calculation differ from a plain `targetEntry->resno`?